src/planner/vectorfield/VectorFieldUtil.cpp

```diff
@@ -6,157 +9,189 @@ namespace aikido {
+    const dart::dynamics::BodyNodePtr bodyNode,
+    double jointLimitPadding,
+    const Eigen::VectorXd* jointVelocityLowerLimits,
+    const Eigen::VectorXd* jointVelocityUpperLimits,
+    double stepSize)
 {
   using dart::math::Jacobian;
   using dart::optimizer::Problem;
   using dart::optimizer::Solver;
   using Eigen::VectorXd;
 
   const dart::dynamics::MetaSkeletonPtr skeleton
-      = _stateSpace->getMetaSkeleton();
+      = stateSpace->getMetaSkeleton();
   // Use LBFGS to find joint angles that won't violate the joint limits.
-  const Jacobian jacobian = skeleton->getWorldJacobian(_bodyNode);
+  const Jacobian jacobian = skeleton->getWorldJacobian(bodyNode);
 
   const std::size_t numDofs = skeleton->getNumDofs();
-  const VectorXd positions = skeleton->getPositions();
-  VectorXd initialGuess = skeleton->getVelocities();
-  const VectorXd positionLowerLimits = skeleton->getPositionLowerLimits();
-  const VectorXd positionUpperLimits = skeleton->getPositionUpperLimits();
-  VectorXd velocityLowerLimits = skeleton->getVelocityLowerLimits();
-  VectorXd velocityUpperLimits = skeleton->getVelocityUpperLimits();
 
-  auto currentState = _stateSpace->createState();
-  _stateSpace->convertPositionsToState(positions, currentState);
+  jointVelocity = Eigen::VectorXd::Zero(numDofs);
+  VectorXd positions = skeleton->getPositions();
+  VectorXd initialGuess = skeleton->getVelocities();
+  VectorXd positionLowerLimits = skeleton->getPositionLowerLimits();
+  VectorXd positionUpperLimits = skeleton->getPositionUpperLimits();
+  auto currentState = stateSpace->createState();
+  stateSpace->convertPositionsToState(positions, currentState);
 
-  for (std::size_t i = 0; i < numDofs; ++i)
+  const auto problem = std::make_shared<Problem>(numDofs);
+  if (jointVelocityLowerLimits != nullptr)
   {
-    const double nextPositionLowerBound
-        = positions[i] + _timestep * velocityLowerLimits[i];
-    const double nextPositionUpperBound
-        = positions[i] + _timestep * velocityUpperLimits[i];
-
-    const double paddedPositionLowerLimit = positionLowerLimits[i] + _padding;
-    const double paddedPositionUpperLimit = positionUpperLimits[i] - _padding;
-
-    if (nextPositionLowerBound < paddedPositionLowerLimit)
+    VectorXd velocityLowerLimits = *jointVelocityLowerLimits;
+    for (std::size_t i = 0; i < numDofs; ++i)
     {
-      const double feasibleVelocityLowerLimit
-          = (paddedPositionLowerLimit - positions[i]) / _timestep;
-      velocityLowerLimits[i]
-          = std::max(velocityLowerLimits[i], feasibleVelocityLowerLimit);
+      const double position = positions[i];
+      const double positionLowerLimit = positionLowerLimits[i];
+      const double velocityLowerLimit = velocityLowerLimits[i];
+
+      if (position + stepSize * velocityLowerLimit
+          <= positionLowerLimit + jointLimitPadding)
+      {
+        velocityLowerLimits[i] = 0.0;
+      }
```

Comment:
This check needs to live outside the `if (jointVelocityLowerLimits != nullptr)` block. We want to enforce joint limits even if we don't enforce velocity limits.

The simplest way to implement this might be to initialize `jointVelocityLowerLimits` a vector of `+infinity()` constants if it is not passed as an argument, then proceed as before. You might want to double check with @jslee02 that the optimizer will behave as expected if we do this.

Reply:
I remember that I had tested that setting `infinity()` to bounds will cause `invalid argument` error in calling `NlOpt solver`. The sample code I could find in `dart` uses `HUGE_VAL`. I think we could users decide what the bounds are like and set accordingly. I added back `enforceJointVelocityLimits` to enable/disable joint velocity limits here.